fontbe/src/features/marks.rs

```diff
@@ -76,31 +83,48 @@ impl MarkGroup<'_> {
     }
 }
 
-// a trait to abstract over two very similar builders
+// a trait to abstract over three very similar builders
 trait MarkAttachmentBuilder: Default {
-    fn add_mark(&mut self, gid: GlyphId, group: &MarkGroupName, anchor: fea_rs::compile::Anchor);
-    fn add_base(&mut self, gid: GlyphId, group: &MarkGroupName, anchor: fea_rs::compile::Anchor);
+    fn add_mark(&mut self, gid: GlyphId, group: &GroupName, anchor: FeaAnchor);
+    fn add_base(&mut self, gid: GlyphId, group: &GroupName, anchor: BaseOrLigaAnchors<FeaAnchor>);
 }
 
 impl MarkAttachmentBuilder for MarkToBaseBuilder {
-    fn add_mark(&mut self, gid: GlyphId, group: &MarkGroupName, anchor: fea_rs::compile::Anchor) {
-        //FIXME: precheck groups to ensure no overlap
+    fn add_mark(&mut self, gid: GlyphId, group: &GroupName, anchor: FeaAnchor) {
         let _ = self.insert_mark(gid, group.clone(), anchor);
     }
 
-    fn add_base(&mut self, gid: GlyphId, group: &MarkGroupName, anchor: fea_rs::compile::Anchor) {
-        self.insert_base(gid, group, anchor)
+    fn add_base(&mut self, gid: GlyphId, group: &GroupName, anchor: BaseOrLigaAnchors<FeaAnchor>) {
+        match anchor {
+            BaseOrLigaAnchors::Base(anchor) => self.insert_base(gid, group, anchor),
+            BaseOrLigaAnchors::Liga(_) => unreachable!(),
+        }
     }
 }
 
 impl MarkAttachmentBuilder for MarkToMarkBuilder {
-    fn add_mark(&mut self, gid: GlyphId, group: &MarkGroupName, anchor: fea_rs::compile::Anchor) {
-        //FIXME: precheck groups to ensure no overlap
+    fn add_mark(&mut self, gid: GlyphId, group: &GroupName, anchor: FeaAnchor) {
         let _ = self.insert_mark1(gid, group.clone(), anchor);
     }
 
-    fn add_base(&mut self, gid: GlyphId, group: &MarkGroupName, anchor: fea_rs::compile::Anchor) {
-        self.insert_mark2(gid, group, anchor)
+    fn add_base(&mut self, gid: GlyphId, group: &GroupName, anchor: BaseOrLigaAnchors<FeaAnchor>) {
+        match anchor {
+            BaseOrLigaAnchors::Base(anchor) => self.insert_mark2(gid, group, anchor),
+            BaseOrLigaAnchors::Liga(_) => unreachable!(),
+        }
+    }
+}
+
+impl MarkAttachmentBuilder for MarkToLigBuilder {
+    fn add_mark(&mut self, gid: GlyphId, group: &GroupName, anchor: FeaAnchor) {
+        let _ = self.insert_mark(gid, group.clone(), anchor);
+    }
+
+    fn add_base(&mut self, gid: GlyphId, group: &GroupName, anchors: BaseOrLigaAnchors<FeaAnchor>) {
+        match anchors {
+            BaseOrLigaAnchors::Liga(anchors) => self.insert_lig(gid, group.clone(), anchors),
+            BaseOrLigaAnchors::Base(_) => unreachable!(),
```

Comment:
Provide _at least_ the offending gid and probably the group name as well. I know it's not supposed to happen but if/when it does we'll curse the lack of context provided.

Reply:
see https://github.com/googlefonts/fontc/blob/945604aa6e0c634165c5d30178ca9a674ade7169/fontbe/src/features/marks.rs#L215-L217 for how this is called;


basically you would need to explicitly use the wrong generic type to hit this branch, in which case it will fail immediately and for all inputs. Added a panic message, though.